### PR TITLE
HDDS-5686. Support transfer leadership in raft based ha cluster

### PR DESCRIPTION
**Pull request by Xushaohong**
## What changes were proposed in this pull request?
Background:
In a raft based ha cluster, for some reason, we want to transfer leadership to a specific follower gracefully, now, we have to kill the leader master to transfer leadership, but the leadership could transfer to an unexpected follower, we have to kill the leader master again, this is so tricky.

Usage:
This feature is assumed to utilizing a shell command like ```ozone admin failover  [SCM/OM]  [host/host:port/'random'] ```, with optional IpAddr we can transfer the leadership to the target follower, with 'random' we can transfer leadership to any node except leader node.

The whole procedure is as following:
   *   Pseudo raft client -> get real raft group info -> check address
   *   whether in raft peers -> real raft client -> set priority -> trigger
   *   transferleadership -> new raft leader takes office

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5686

Please replace this section with the link to the Apache JIRA)

## How was this patch tested?
manual tests. 
the Integration test will be added soon

Here are some screenshots
截屏2021-09-28 下午3 59 40
截屏2021-09-28 下午3 48 29
截屏2021-09-28 下午3 51 06

### CODE REVIEW

**Comment by Xushaohong** [comment]
@bharatviswa504 @bshashikant PTAL, THANKS!

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/failover/FailoverCommand.java [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+  @CommandLine.Option(
+      names = {"-id", "--service-id"},
+      description = "Ozone Manager Service ID, if domain is om, " +
+          "this option is prerequisite"
+  )
+  private String omServiceId;
+
+  @CommandLine.Option(
+          names = {"--raftHostPortList"},
```
Comment (bshashikant) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/failover/FailoverCommand.java:97]: can we replace raft host with generic "host"

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/failover/FailoverCommand.java [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+  private String domain;
+
+  @CommandLine.Parameters(
+      description = "[host/host:port/'random'] ratis rpc address of " +
+      "target domain if 'random' then choose arbitrary follower node "
+  )
+  private String tgtAddress;
+
+  @CommandLine.Option(
+      names = {"-id", "--service-id"},
```
Comment (bshashikant) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/failover/FailoverCommand.java:90]: do we really need to give the  port? . if its always going to be read from the configuration, proablyt , we can omit this option.

**Review by bshashikant — COMMENTED** [review]
i am not sure, if the cmd does any authetication/security validation here. The raft client is directly talking to the raft server. We need to fit the cmd to security model in ozone.

cc  @bharatviswa504

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/failover/FailoverCommand.java (hunk 2 above)
Comment (Xushaohong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/failover/FailoverCommand.java:90]: This is for the case where SCMs have consistent but different ports from the default. This key ```OZONE_SCM_RATIS_PORT_KEY``` is only used between SCM nodes right now and the config of the client may not cover it.
Another scenario could be that SCMs have inconsistent ratis ports such as MiniOzoneCluster.

**Comment by Xushaohong** [comment]
> i am not sure, if the cmd does any authetication/security validation here. The raft client is directly talking to the raft server. We need to fit the cmd to security model in ozone.
> 
> cc @bharatviswa504

This implementation looks a  little tricky, it's indeed directly talking to the raft server. I agree that authentication/security should be added in the future.  : )

**Comment by bharatviswa504** [comment]
> > i am not sure, if the cmd does any authetication/security validation here. The raft client is directly talking to the raft server. We need to fit the cmd to security model in ozone.
> > cc @bharatviswa504
> 
> This implementation looks a little tricky, it's indeed directly talking to the raft server. I agree that authentication/security should be added in the future. : )

One idea might be similar to DN, contact OM and get list of CA and use that in setting up with raft client. And we can make this an admin command, and in secure cluster we need kerberos ticket to run this command. 

For security question this way we can solve it. And also we need this security/authorization implemented, otherwise this command cannot work in a secure cluster.

Sample code:
```
    final GrpcTlsConfig tlsConfig = RatisHelper.createTlsClientConfig(new
        SecurityConfig(ozoneConf), caCerts);
```
And use this tlsConfig when creating RaftClient. Here caCerts is obtained from OM getServiceList Call. You can refer RpcClient for this.

**Comment by bshashikant** [comment]
@bharatviswa504 , can you please check this one?

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/GetCertificatesSubcommand.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      names = {"-host", "--service-host"},
+      description = "Ozone Manager Host, for non-HA mode."
+  )
+  private String omHost;
+
+  @CommandLine.Option(
+      names = {"--show", "-show"},
+      required = true
+  )
+  private boolean action;
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/GetCertificatesSubcommand.java:60]: This option is not used.  What is the purpose?

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/GetCertificatesSubcommand.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * Handler of om roles command.
+ */
+@CommandLine.Command(
+    name = "certificates", aliases = "certs",
+    description = "List all CA certificates from OM",
+    mixinStandardHelpOptions = true,
+    versionProvider = HddsVersionProvider.class)
+public class GetCertificatesSubcommand implements Callable<Void> {
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/GetCertificatesSubcommand.java:40]: This subcommand does not seem to be necessary for the failover.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/GetCertificatesSubcommand.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  @CommandLine.ParentCommand
+  private OMAdmin parent;
+
+  @CommandLine.Option(names = {"-id", "--service-id"},
+      description = "OM Service ID, for HA mode."
+  )
+  private String omServiceId;
+
+  @CommandLine.Option(
+      names = {"-host", "--service-host"},
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/GetCertificatesSubcommand.java:51]: Single-dash options should be single-char (e.g. `-h`), not longer (`-host`).  If the char is already used for other option, consider capital (`-H`), another char or omitting the single-dash option completely.

Rationale: different single-dash options could be combined in the same param, e.g. `-host` is expected to be the same as `-h -o -s -t`.

This applies to all options, not just to this specific one.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/failover/FailoverCommand.java [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+      configuration = new OzoneConfiguration();
+    } else{
+      configuration = parent.getOzoneConf();
+    }
+    try {
+      transferLeadership(configuration);
+    } catch (Exception ex) {
+      ex.printStackTrace();
+      throw ex;
+    }
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/failover/FailoverCommand.java:129]: Ozone CLI handles uncaught exception and prints message or stack trace depending on this global option:

```
--verbose   More verbose output. Show the stack trace of the errors.
```

So the `printStackTrace()` is unnecessary here.  (And then `try-catch` can be omitted, too.)

```suggestion
    transferLeadership(configuration);
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @Xushaohong for working on this.  The basic idea seems good.

In addition to the specific comments below, I find the command's output to be a bit too verbose and eclectic.  I think most of the current output should be printed only with the `--verbose` flag or at debug level.

**Comment by ChenSammi** [comment]
Hey @Xushaohong , could you update the patch according to our discussion at your convenient time?

**Comment by Xushaohong** [comment]
> Hey @Xushaohong , could you update the patch according to our discussion at your convenient time?

Sure， i will open a new PR ASAP

**Pull request closed without merging**